numpy/lib/tests/test_function_base.py

```diff
@@ -2354,14 +2365,72 @@ def test_scalar_interpolation_point(self):
         x0 = np.nan
         assert_almost_equal(np.interp(x0, x, y), x0)
 
-    def test_non_finite_behavior(self):
+    def test_non_finite_behavior_exact_x(self):
         x = [1, 2, 2.5, 3, 4]
         xp = [1, 2, 3, 4]
         fp = [1, 2, np.inf, 4]
         assert_almost_equal(np.interp(x, xp, fp), [1, 2, np.inf, np.inf, 4])
         fp = [1, 2, np.nan, 4]
         assert_almost_equal(np.interp(x, xp, fp), [1, 2, np.nan, np.nan, 4])
 
+    @pytest.fixture(params=[
+        lambda x: np.float_(x),
+        lambda x: _make_complex(x, 0),
+        lambda x: _make_complex(0, x),
+        lambda x: _make_complex(x, np.multiply(x, -2))
+    ], ids=[
+        'real',
+        'complex-real',
+        'complex-imag',
+        'complex-both'
+    ])
+    def sc(self, request):
+        """ scale function used by the below tests """
+        return request.param
+
+    def test_non_finite_any_nan(self, sc):
+        """ test that nans are propagated """
+        assert_equal(np.interp(0.5, [np.nan,      1], sc([     0,     10])), sc(np.nan))
+        assert_equal(np.interp(0.5, [     0, np.nan], sc([     0,     10])), sc(np.nan))
+        assert_equal(np.interp(0.5, [     0,      1], sc([np.nan,     10])), sc(np.nan))
+        assert_equal(np.interp(0.5, [     0,      1], sc([     0, np.nan])), sc(np.nan))
+
+    def test_non_finite_inf(self, sc):
+        """ Test that interp between opposite infs gives nan """
+        assert_equal(np.interp(0.5, [-np.inf, +np.inf], sc([      0,      10])), sc(np.nan))
+        assert_equal(np.interp(0.5, [      0,       1], sc([-np.inf, +np.inf])), sc(np.nan))
+        assert_equal(np.interp(0.5, [      0,       1], sc([+np.inf, -np.inf])), sc(np.nan))
+
+        # unless the y values are equal
+        assert_equal(np.interp(0.5, [-np.inf, +np.inf], sc([     10,      10])), sc(10))
```

Comment:
This test is the new one

Reply:
That was similar to some of the things I was considering, but overall I am not sure I am too worried. At some point floating point ranges get wonky anyway. E.g. as soon as the distance of both sides are infinite (and not exactly one of them is finite), things get awkward unless the numbers are identical. I do not think there is a way around this, so I was starting to think there are simply some things we should accept as floating point precision errors.

Reply:
> things get awkward unless the numbers are identical.

Correct. My original strategy was to try and reason about this awkwardness, which missed this case, and probably some others.

The new strategy is based on the assumption that floating point math will never give a finite result if the operation is not well defined, and so there's no harm in trying a bunch of different formulations until we find a finite one.

Reply:
Yeah, this is good, thanks for keeping poking it. If there are remaining holes, I do not think they can be avoided, and these changes make all things with infinities correct as far as I can see.

Should we short release notes? I can do that as well though.

Reply:
Release notes wouldn't be a bad idea. "better handling of non-finite values in `interp`" or something under improvements.

I'm not really sure whether to consider this a bug fix or an enhancement.

Reply:
Hmm, good point. That sounds good for the notes. Now that you are asking, my gut feeling is a slight preference for enhancement.